### ISSUE TITLE
Support for Plugins without a parameter less constructor

### ISSUE THREAD

**Issue by ghost** [jordimontana82/fake-xrm-easy#341]
Hello @jordimontana82 
I am having a problem with a plugin which doesn't have a default parameter less constructor but only a constructor like `ExamplePlugin(string unsecure, string secure)` and I can't call `ExecutePluginWith`. 
I think that it would be nice to have overloads for `IXrmContext ` methods which don't use generics to support plugins/workflows which don't have a parameter less constructor, for example, each method that needs an plugin instance to have the existing method `IPlugin ExecutePluginWith<T>(XrmFakedPluginExecutionContext ctx, T instance) where T : IPlugin, new();
` and as simpler method like `IPlugin ExecutePluginWith(XrmFakedPluginExecutionContext ctx, IPlugin instance);` because actually in this case you already have the plugin instance and doesn't make sense to use the Generic Type constructor, because we already are providing the instance of the plugin.

**Comment by jordimontana82** [comment]
Hi @BetimBeja,

Thanks for raising this. You should be able to use [ExecutePluginWithConfigurations](https://github.com/jordimontana82/fake-xrm-easy/blob/b05fffaffb9fa86aeda0abb969677c4632b8154f/FakeXrmEasy.Shared/IXrmFakedContext.cs#L108-L113) instead, which also comes with 2 overloads, one of those also accepting a specific instance.

As there is no other way Dynamics 365 CE / CRM creates instances for plugins (either without parameters or with configurations) we're in theory already covered and we don't need extra overloads.

Hope this helps

**Comment by jordimontana82** [comment]
Hi @BetimBeja 

Good point, yep the ExecutionPluginWithConfiguration method that receives the instance doesn't need the config parameters, we could just remove them and just use the existing ExecutePluginWith<> along with removing the type, indeed, which is really needed only for methods that take no instances as parameters and so the type is needed to be able to create the instance inside the method.

Re the other methods, apart from the ExecutePluginWithTarget, I rarely used them anymore since we have the ability to pass the whole custom execution context with the properties we need. I'm talking about methods like ExecutePluginWithTargetAndPreEntityImages, etc, these existed before the XrmFakedPluginContext was introduced, which was introduced mainly because as soon as you wanted to pass many context properties the method call would end up being pretty long, and not very intuititive.

So, my point is, yeah, we could also probably introduce the instance parameter to ExecutePluginWithTarget and (probably now that I'm checking ExecutePluginWithTargetReference) but I wouldn't do it for the others (ExecutePluginWithTargetAndInputParameters, ExecutePluginWithTargetAndPostEntityImages, ExecutePluginWithTargetAndInputParameters, ..) which don't have that yet and we should probably mark them as deprecated and better use the  full plugin context methods in those cases.

That's just my preference :)

Thoughts?

**Comment by BetimBeja** [comment]
@jordimontana82 
I am a little new to this library and i opened the issue because i think this is a problem anyone new might encounter. 
I agree the deprecation of legacy function is a good thing, as long as it is done with at least one release were it is marked with an Obsolete attribute. Also i think that ExecutionPluginWithConfiguration overload which accepts an instance should be marked as Obsolete suggesting the use of ExecutePluginWith before removal.
Also an off topic suggestion would be the inclusion of an Contribute article on the wiki (where you can suggest a typical development environment) so anyone interested in contributing to the project can easily get up and running, personally i am having some problems with building the master branch from the repository. 
Keep up with the good work!

**Comment by BetimBeja** [comment]
Hello,
Since #343 was merged i think this issue should be closed!

Best regards,
@BetimBeja

**Comment by jordimontana82** [comment]
Hi @BetimBeja 

Indeed, sorry.

That's closed now. Another option would be to add a commit message like "Closes #341" and it will automatically close it too once Pull request is merged :)

**Comment by BetimBeja** [comment]
Hi @jordimontana82 ,
I think that i will get better at this with the next commits :) 
i think that issue #331 has the same problem as this since #339 fixes it.

Best regards,
Betim.

**Comment by jordimontana82** [comment]
Same thing yeah :D
